GAE/src/org/akvo/flow/dao/DataPointAssignmentDao.java

```diff
@@ -0,0 +1,74 @@
+/*
+ *  Copyright (C) 2019 Stichting Akvo (Akvo Foundation)
+ *
+ *  This file is part of Akvo FLOW.
+ *
+ *  Akvo FLOW is free software: you can redistribute it and modify it under the terms of
+ *  the GNU Affero General Public License (AGPL) as published by the Free Software Foundation,
+ *  either version 3 of the License or any later version.
+ *
+ *  Akvo FLOW is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ *  without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *  See the GNU Affero General Public License included below for more details.
+ *
+ *  The full license text can also be seen at <http://www.gnu.org/licenses/agpl.html>.
+ */
+
+package org.akvo.flow.dao;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.jdo.PersistenceManager;
+
+import org.akvo.flow.domain.persistent.DataPointAssignment;
+import org.akvo.flow.domain.persistent.SurveyAssignment;
+
+import com.gallatinsystems.framework.dao.BaseDAO;
+import com.gallatinsystems.framework.servlet.PersistenceFilter;
+import com.gallatinsystems.user.domain.User;
+
+public class DataPointAssignmentDao extends BaseDAO<DataPointAssignment> {
+
+    public DataPointAssignmentDao() {
+        super(DataPointAssignment.class);
+    }
+
+    /**
+     * Return a set of data point ids for a specified device and survey (in any assignment)
+     *
+     * @return list of assignments
+     */
+    public Set<Long> listDataPointIds(Long deviceId, Long surveyId) {
+        SurveyAssignmentDao saDao = new SurveyAssignmentDao();
+        Set<Long> result = new HashSet<>();
+        List<SurveyAssignment> assignments = saDao.listBySurveyGroup(surveyId);
+        for (SurveyAssignment sa: assignments) {
```

Comment:
Its usually considered an antipattern for the datastore to have a query within a loop. It causes too much chatter between the datastore and the backend. Would it be possible to modify this to be say `listDataPointIds(Long deviceId, Long surveyAssignmentId)`?

Reply:
Then the loop just moves up to DataPointRestService, which uses this method.

Reply:
The only way I can see to do it without loops is to normalise totally; have an entity for each (datapoint,device,assignment) triple and select on device only. But there will be thousands of them...

Reply:
I guess my assumption was that there would be very few datapointassignments with the same device+surveyassigment

Reply:
Doing it your way.